internal/internal_workflow_client.go

```diff
@@ -734,6 +735,55 @@ func (wc *WorkflowClient) QueryWorkflow(ctx context.Context, workflowID string,
 	})
 }
 
+// UpdateWorkflowWithOptionsRequest is the request to UpdateWorkflowWithOptions
+type UpdateWorkflowWithOptionsRequest struct {
+	// UpdateID is an application-layer identifier for the requested update. It
+	// must be unique within the scope of a Namespace+WorkflowID+RunID.
+	UpdateID string
+
+	// WorkflowID is a required field indicating the workflow which should be
+	// updated.
+	WorkflowID string
+
+	// RunID is an optional field used to identify a specific run of the target
+	// workflow.  If RunID is not provided the latest run will be used.
+	RunID string
+
+	// UpdateName is a required field which specifies the update you want to run.
+	// See comments at workflow.SetUpdateHandler(ctx Context, updateName string, handler interface{}, opts UpdateHandlerOptions)
+	// for more details on how to setup update handlers within the target workflow.
+	UpdateName string
+
+	// Args is an optional field used to identify the arguments passed to the
+	// update.
+	Args []interface{}
+
+	// FirstExecutionRunID specifies the RunID expected to identify the first
+	// run in the workflow execution chain. If this expectation does not match
+	// then the server will reject the update request with an error.
+	FirstExecutionRunID string
+
+	// this isn't upstream in API yet
+	// WaitFor enumspb.WorkflowExecutionUpdateWaitEvent
+}
+
+// WorkflowUpdateHandle is a handle to a workflow execution update
+// process. The update may or may not have completed so an instance of this type
+// functions simlarly to a Future with respect to the result value of the
+// update. If the update is rejected or returns an error, the Get function on
+// this type will return that error.
+type WorkflowUpdateHandle interface {
```

Comment:
I don't think we're ready to require Go 1.18, are we? @cretz ?

Reply:
I mean, I think we are, but that's unrelated to @bergundy's point. He means "generic" I think in the English sense not the programmer sense :-)

@bergundy - I don't think this needs to be a generic future. There will be more update-specific calls here later for doing things like describing the update. We can abstract to a shared future interface later if there is value, but there is no value in doing it now. Do you have a specific reason for a generic future besides that it just feels right in your OO mind?

Reply:
I can see more methods being added here when we add an ability to get a handle for an existing update.
We'll probably want to be able to wait for acceptance for example.
Describing an update seems reasonable to me too TBH but probably not as useful on handles returned by the update-initiating APIs.

For the current functionality, I don't see a justification in returning the specialized handle.
My reason is more YAGNI (for this flow) than "doesn't feel right".

Reply:
>  My reason is more YAGNI (for this flow) than "doesn't feel right".

"You Aren't Gonna Need It" where "It" = "generic future abstraction". The describing of an update is likely gonna be needed as is cancelling, a getter for workflow ID, etc. `WorkflowRun` isn't a generic future either. And since Go is duck typed, you can type this in your code:

```go
type Future interface {
  Get(context.Context, ...interface{}) error
}
```

And now you have a generic future for `WorkflowUpdateHandle` and `WorkflowRun`. But it's silly to make some generic thing in a non-generic situation. This is a update handle very simply.

Reply:
We need a getter for ID and WFRunID (because it's otherwise not knowable if it's not specified at invocation time). That's why there's a getter commented out on the line immediately under the line we're all commenting on. If the type was available in API today it would be included in this PR. It is known and understood to be required to re-hydrate update handles. We will need it to ship even a pre-release.

Reply:
:+1: We could go ahead and add `GetID()` right now since you have it, but yeah no use in it except for logging yet until we have `GetUpdate` or whatever in the future.